src/stransport_stream.c

```diff
@@ -150,9 +156,9 @@ static OSStatus read_cb(SSLConnectionRef conn, void *data, size_t *len)
 	requested = left = *len;
 	do {
 		ret = git_stream_read(io, data + (requested - left), left);
-		if (ret < 0) {
+		if (ret <= 0) { /* There was an error or nothing could be read because stream is closed */
 			*len = 0;
-			return -1;
+			return errSSLClosedAbort;
```

Comment:
Originally, I modified this function to behave like "read up to *len bytes from the connection" i.e. get rid of the `do ... while`. It worked the same way than the write function: "write up to *len bytes to the connection". In both cases if fewer bytes were handled, it would still return `noErr`.

But my functional tests failed when cloning a repo over HTTPS. So I figured, even though the docs do not mention that, maybe the behavior should be: "read or write this *exact* number of bytes to the connection or fail entirely if it can't". That's what you had done initially (why BTW?). I restored the `do ... while` in the read function and added the same behavior to the write function.

If you assume that's the desired behavior, and considering 0 can only be returned if fewer than `*len` were read, then the function should return an error. I picked `errSSLClosedAbort` for both read and write functions as it seemed the most sensitive among the possible SSL error codes.